src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -259,417 +258,482 @@ AND NOT EXISTS (
-        // outer-most query. We can't use {alias} here, since we're basing the subquery on Product
-        // instead of Pool to save ourselves an unnecessary join. Similarly, we use an SQL
-        // restriction here because we can query the information we need, hitting only one table
-        // with direct SQL, whereas the matching criteria query would end up requiring a minimum of
-        // one join to get from pool to pool attributes.
-        DetachedCriteria prodAttrSubquery = DetachedCriteria.forClass(Product.class, "ProdI")
-            .createAlias("ProdI.attributes", "attrib")
-            .setProjection(Projections.id())
-            .add(Property.forName("Product.uuid").eqProperty("ProdI.uuid"))
-            .add(Restrictions.eq("attrib.indices", key))
-            .add(Restrictions.sqlRestriction(
-                "NOT EXISTS (SELECT poolattr.pool_id FROM cp_pool_attribute poolattr " +
-                    "WHERE poolattr.pool_id = this_.id AND poolattr.name = ?)",
-                key, StringType.INSTANCE));
-
-        if (values != null && !values.isEmpty()) {
-            Disjunction poolAttrValueDisjunction = Restrictions.disjunction();
-            Disjunction prodAttrValueDisjunction = Restrictions.disjunction();
-
-            for (String attrValue : values) {
-                if (attrValue == null || attrValue.isEmpty()) {
-                    poolAttrValueDisjunction.add(Restrictions.isNull("attrib.elements"))
-                        .add(Restrictions.eq("attrib.elements", ""));
-
-                    prodAttrValueDisjunction.add(Restrictions.isNull("attrib.elements"))
-                        .add(Restrictions.eq("attrib.elements", ""));
-                }
-                else {
-                    attrValue = this.sanitizeMatchesFilter(attrValue);
-                    poolAttrValueDisjunction.add(CPRestrictions.ilike("attrib.elements", attrValue, '!'));
-                    prodAttrValueDisjunction.add(CPRestrictions.ilike("attrib.elements", attrValue, '!'));
-                }
+            Consumer host = this.consumerCurator.getHost(uuidFact, ownerId);
+            if (host == null) {
+                return null;
             }
 
-            poolAttrSubquery.add(poolAttrValueDisjunction);
-            prodAttrSubquery.add(prodAttrValueDisjunction);
+            // Impl note:
+            // This query matches pools with the "requires_host" attribute explicitly set to a
+            // value other than the host we're looking for. We then negate the results of this
+            // subquery, so our final result is: fetch pools which do not have a required host
+            // or have a required host equal to our host.
+
+            Subquery<String> subQuery = query.subquery(String.class);
+            Root<Pool> subQueryRoot = subQuery.from(Pool.class);
+            subQuery.select(subQueryRoot.get(Pool_.id));
+
+            CriteriaBuilder subQueryCB = this.entityManager.get().getCriteriaBuilder();
```

Comment:
Is this necessary? I'm pretty sure you can just use the builder provided to the method.